### ISSUE TITLE
[feature request] RNN Wishlist

### ISSUE THREAD

**Issue by Kaixhin** [pytorch/pytorch#9572]
PT has fairly basic options for RNNs, mainly due to keeping parity with cuDNN - something that should be alleviated by 1.0, which would allow a more flexible RNN API (https://github.com/pytorch/pytorch/issues/711). This issue details some possible options that we may want to provide. Sonnet's [gated_rnn file](https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/gated_rnn.py) provides a nice reference for all of these.

- [ ] Customisable activation functions (https://github.com/pytorch/pytorch/issues/1173). Would add a lot of flexibility, and needed to reproduce some works (see linked issue).
- [ ] Layer normalisation built in (https://github.com/pytorch/pytorch/issues/6760). The [original paper](https://arxiv.org/pdf/1607.06450.pdf) provides 2 LSTM variants, and 1 GRU variant. TF provides a [different module](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LayerNormBasicLSTMCell) for this. Pretty widely used due to its ability to improve stability in training, among other possible benefits.
- [ ] Recurrent dropout (https://github.com/pytorch/pytorch/issues/2166). An RNN regulariser - see [original paper](https://arxiv.org/pdf/1603.05118.pdf). TF provides this in the [layer norm LSTM](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LayerNormBasicLSTMCell), but it could be applied as a generic wrapper for RNNs - see [Sonnet](https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/gated_rnn.py) for example.
- [ ] Zoneout (https://github.com/pytorch/pytorch/issues/561). [Original paper](https://arxiv.org/pdf/1606.01305.pdf). This is again a generic RNN regulariser, implemented by a wrapper in [Sonnet](https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/gated_rnn.py).
- [ ] Configurable forget gate bias for LSTM (https://github.com/pytorch/pytorch/issues/750). Setting this to 1 is advocated for by the [original forget gate paper](https://www.researchgate.net/profile/Felix_Gers/publication/12292425_Learning_to_Forget_Continual_Prediction_with_LSTM/links/5759414608ae9a9c954e84c5/Learning-to-Forget-Continual-Prediction-with-LSTM.pdf) - this makes the LSTM remember information until it learns to forget. Noticeably, this is the default in [TF](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) and [Keras](https://keras.io/layers/recurrent/#lstmcell).
- [ ] Peephole connections (https://github.com/pytorch/pytorch/issues/630). According to the [original paper](http://www.jmlr.org/papers/volume3/gers02a/gers02a.pdf), can help learn certain sequences.
- [ ] LSTM cell clipping. Provided in [TF](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell), not on by default. [Source](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43905.pdf) via @ngimel.
- [ ] Projection layer in LSTM. Allows larger units - see [original paper](https://arxiv.org/pdf/1402.1128.pdf) - that are particularly relevant in NLP tasks. Provided in [TF](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell).
- [ ] Convolutional LSTM (https://github.com/pytorch/pytorch/issues/1706). [Original paper](https://arxiv.org/pdf/1506.04214.pdf) and [TF module](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/ConvLSTMCell) - widely used component by now.
- [ ] Gradient clipping at every timestep (see comments from @PetrochukM below for details). [Original paper](https://arxiv.org/pdf/1211.5063.pdf).
- [ ] Better initialisations? Keras uses [Glorot uniform for input-to-hidden, orthogonal for hidden-to-hidden weights, and zeros for the bias](https://keras.io/layers/recurrent/), across all unit types (simple RNN, LSTM and GRU). This fits into the larger question of highlighting better weight initialisations across all PT modules, so may be out of place here.
- [ ] Single bias? The current RNNs have two biases, which is completely redundant.

**Comment by t-vi** [comment]
I have a pure-python-PyTorch LSTM where I expanded the one from pytorch/benchmark to reflect all current options (plus some dropout). If that can be of use for making a reference to test against, I'd be happy to put that up somewhere.

**Comment by Kaixhin** [comment]
@t-vi if you've already got a Python PyTorch version of these that you can share then posting a link would be useful, thanks.

**Comment by ngimel** [comment]
cudnn supports projection layer in LSTM starting from 7.1.2, it is not exposed in pytorch though, as we did not see a lot of demand for it. cudnn will support LSTM cell clipping in the next version.

**Comment by PetrochukM** [comment]
- [ ] To help with gradient explosions, gradient clipping at every step of the RNN. As suggested in the [original paper](https://arxiv.org/abs/1211.5063), that introduced gradient clipping. Difficult to train RNNs on long sequences otherwise without get a NaN.

**Comment by Kaixhin** [comment]
@ngimel Do you have a reference for where cell clipping originated from? I've been trying to track down its source but haven't turned up anything so far.

@PetrochukM Is this something different to [`clip_grad_norm_`](https://pytorch.org/docs/stable/nn.html?highlight=clip#torch.nn.utils.clip_grad_norm_) or [`clip_grad_value_`](https://pytorch.org/docs/stable/nn.html?highlight=clip#torch.nn.utils.clip_grad_value_)?

**Comment by PetrochukM** [comment]
@Kaixhin Yeah it is. Those functions operate after the entire gradient has been computed. 

That’s fine if the sequence is short but with long sequences the gradient could explode to infinity before it’s clipped.

**Comment by Kaixhin** [comment]
**Update from the team:** there's ongoing work to make sure that vanilla PyTorch RNNs (with for loops) hit cuDNN speeds when JITted: https://github.com/pytorch/benchmark/tree/master/rnns . More information about this will be provided with the 1.1 release, when these changes will be incorporated.

**Comment by Kaixhin** [comment]
@t-vi with Pytorch 1.1 now out, what do you think is the best way forward on these?

**Comment by t-vi** [comment]
JIT is definitely the way to go here (see the PyTorch blog), even if the backward optimal is still waiting to be merged.

**Comment by soumith** [comment]
I'm going to close this issue, referencing the blog post https://pytorch.org/blog/optimizing-cuda-rnn-with-torchscript/


Custom RNNs via JIT are the way forward.

**Comment by Kaixhin** [comment]
I would say the goal is to provide these out-of-the-box for end users. JIT is great for custom methods, but right now anyone who wants to use layer norm in an RNN (really common use-case) will have to copy over a reference implementation for their own projects. @t-vi is currently drafting a lot of the features for most of these [here](https://github.com/t-vi/pytorch-tvmisc/blob/master/rnn/jited_rnns.ipynb) - hope it can be packaged up as part of torch or at least contrib?

**Comment by soumith** [comment]
yes, we can package them in contrib for sure. right now they are shipped in the `benchmarks` repo, which is probably not as accessible.

**Comment by sushmit86** [comment]
Can someone point out if the recurrent_dropout was implemented or not?
